lib/mpool/lib/mblock_file.c

```diff
@@ -997,6 +1035,65 @@ mblock_file_delete(struct mblock_file *mbfp, uint64_t *mbidv, int mbidc)
     return 0;
 }
 
+merr_t
+mblock_punch(struct mblock_file *mbfp, uint64_t mbid, off_t off, size_t len)
+{
+    struct mblock_oid_info mbinfo;
+    uint32_t block;
+    bool prealloced;
+    size_t wlen;
+    merr_t err;
+    char *addr;
+    int rc;
+
+    INVARIANT(mbfp);
+
+    if (!PAGE_ALIGNED(off) || !PAGE_ALIGNED(len))
+        return merr(EINVAL);
+
+    wlen = mblock_wlen_get(mbfp, mbid);
+
+    if (off < 0 || off >= wlen || off + len > wlen)
+        return merr(EINVAL);
+
+    prealloced = mblock_is_prealloced(mbfp, mbid);
+    if (len == 0 || off + len == wlen) {
+        len = prealloced ? mbfp->mblocksz - off : wlen - off;
+        assert(PAGE_ALIGNED(len));
+    }
+
+    block = block_id(mbid);
+    err = mblock_rgn_find(&mbfp->rgnmap, block + 1);
+    if (err)
+        return err;
+
+    rc = fallocate(mbfp->fd, FALLOC_FL_PUNCH_HOLE | FALLOC_FL_KEEP_SIZE,
+                   off + block_off(mbid, mbfp->mblocksz), len);
+    if (rc == -1)
+        return merr(errno);
+
+    /* Adjust wlen if the punch range includes the end of an mblock */
+    if (off + len >= wlen) {
+        wlen = off;
+        prealloced = false;
+    }
+
+    mblock_wlen_set(mbfp, mbid, wlen, prealloced, true);
```

Comment:
I'm confused about why `wlen` is updated here (and it seems in many other places) but not `alen`.

Reply:
mpool keeps track of the wlen property both in-memory and on-media.
alen is calculated on the fly, please see mblock_file_find().